File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/ApiRequestString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class ApiRequestString {
    private static final String DNCString = "?DNC?";

    public String method;
    public String page;
    public String parameter;
    public String body;

    public ApiRequestString(String method, String page, String parameter, String body) {
        this.method = method;
        this.page = page;
        this.parameter = parameter;
        this.body = body;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static ApiRequestString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 4, "ApiRequest");
        return new ApiRequestString(parts.get(0), parts.get(1), parts.get(2), parts.get(3));
    }

    public ApiRequestString(String text) {
        ApiRequestString parsed = fromText(text);
        this.method = parsed.method;
        this.page = parsed.page;
        this.parameter = parsed.parameter;
        this.body = parsed.body;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof ApiRequestString)) return false;
        ApiRequestString that = (ApiRequestString) o;
        return (DNCString.equals(method) || DNCString.equals(that.method) || Objects.equals(method, that.method))
            && (DNCString.equals(page) || DNCString.equals(that.page) || Objects.equals(page, that.page))
            && (DNCString.equals(parameter) || DNCString.equals(that.parameter) || Objects.equals(parameter, that.parameter))
            && (DNCString.equals(body) || DNCString.equals(that.body) || Objects.equals(body, that.body));
    }

    @Override
    public int hashCode() {
        return Objects.hash(method, page, parameter, body);
    }

    @Override
    public String toString() {
        return Tokens.token(method) + " " + Tokens.token(page) + " " + Tokens.token(parameter) + " " + Tokens.token(body);
    }
}

File: JavaScript/ExampleTests/common/ApiRequestString.js
import * as tokens from "./tokens.js";

export class ApiRequestString {
  static DNC_STRING = "?DNC?";

  constructor(method = "", page = "", parameter = "", body = "") {
    this.method = method;
    this.page = page;
    this.parameter = parameter;
    this.body = body;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new ApiRequestString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? "",
      v[3] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 4, "ApiRequest");
    return new ApiRequestString(
      parts[0],
      parts[1],
      parts[2],
      parts[3]
    );
  }

  toString() {
    return tokens.token(this.method) + " " + tokens.token(this.page) + " " + tokens.token(this.parameter) + " " + tokens.token(this.body);
  }

  equals(other) {
    if (!(other instanceof ApiRequestString)) return false;
    return (this.method === ApiRequestString.DNC_STRING || other.method === ApiRequestString.DNC_STRING || this.method === other.method)
      && (this.page === ApiRequestString.DNC_STRING || other.page === ApiRequestString.DNC_STRING || this.page === other.page)
      && (this.parameter === ApiRequestString.DNC_STRING || other.parameter === ApiRequestString.DNC_STRING || this.parameter === other.parameter)
      && (this.body === ApiRequestString.DNC_STRING || other.body === ApiRequestString.DNC_STRING || this.body === other.body);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}
